### DISCUSSION TITLE
build my copy-paste with windows

### DISCUSSION THREAD

**Discussion by foxnight2** [paullouisageneau/libdatachannel#1053] [Q&A]
I try to generate the datachannel library and use the CMakeList.cmake

```
cmake_minimum_required(VERSION 3.7)

add_executable(datachannel-copy-paste-offerer offerer.cpp)
set_target_properties(datachannel-copy-paste-offerer PROPERTIES
	CXX_STANDARD 17
	OUTPUT_NAME offerer)
target_link_libraries(datachannel-copy-paste-offerer ${DataChannel_LIBRARIES} )
target_include_directories(datachannel-copy-paste-offerer PUBLIC ${DataChannel_INCLUDE_DIRS})

```

Here is my library loacantion
![image](https://github.com/paullouisageneau/libdatachannel/assets/142980812/de2fdedd-d0e1-481c-8a9f-2556eeadd82c)

But the runtime offerer.exe is lack of datachannel.dll

How can I do ?

**Comment by chobie** [comment]
Are you running `cmake` in the top directory of libdatachannel?

If you run `cmake` and then `nmake` as per the following link, I believe samples should be created in the `libdatachannel/build` directory.

https://github.com/paullouisageneau/libdatachannel/blob/master/BUILDING.md#microsoft-windows-with-mingw-cross-compilation

```
$ cmake -B build -G "NMake Makefiles"
$ cd build
$ nmake
```

<img width="662" alt="copy-paste" src="https://github.com/paullouisageneau/libdatachannel/assets/195038/a31ed2b0-57ac-4ff9-9f2e-2b3547b3c4e3">

**Comment by foxnight2** [comment]
No, I am not running cmake in the top directory of libdatachannel.
I just only use or reference https://github.com/paullouisageneau/libdatachannel/blob/master/examples/copy-paste/CMakeLists.txt (this cmake)

Because I already build the libdatachannel static or dynamic  library.
I want to let copy-paste using my build libdatachannel to run.

**Comment by chobie** [comment]
Understood, to summarize, what you want to do is to create a build using `libdatachannel` in your own program, right?

Although it's not a direct sample, I think the `cmake` for obs-webrtc can be a reference for you.
https://github.com/obsproject/obs-studio/blob/93f5b45be8c8b91199c52cf7a0ab59d8c3a08760/plugins/obs-webrtc/CMakeLists.txt#L11

For information on the behavior of `find_package`, please refer to the following link. You will probably need to write something like `libdatachannel-config.cmake`.

https://cmake.org/cmake/help/latest/command/find_package.html

When using `find_package`, you will need a file like this:

https://github.com/paullouisageneau/libdatachannel/blob/cafd410f190dd98e4802ce3ec5e864b72ffd7755/cmake/Modules/FindLibJuice.cmake

The `copy-paste/CMakeLists.txt` is set in the parent cmake settings and is configured to reference `libdatachannel`, so it might be difficult to use it as a reference.

**Comment by chobie** [comment]
This is a discussion about the basic usage of cmake, so I think it would be better to first learn how to use cmake properly.
I only write it occasionally, so I often forget how to use it...

Alternatively, if you are only building for Windows, it might be easier to set up in Visual Studio without using cmake.

**Comment by paullouisageneau** [comment]
CMake can't guess that libdatachannel is installed in `C:\tmp\prebuild\libdatachannel`, you need to add the directory with `-DCMAKE_PREFIX_PATH="C:\tmp\prebuild\libdatachannel"` (see the [documentation](https://cmake.org/cmake/help/v3.28/variable/CMAKE_PREFIX_PATH.html)).

What you are trying to do is typically cumbersome on Windows as Windows does not have standard system-wide include and lib directories to install the libraries. I'd recommend putting the libdatachannel source inside your project instead and including it with `add_subdirectory(libdatachannel)`, then you can directly use `target_link_libraries(your-project LibDataChannel::LibDataChannel)`.

As @chobie already mentionned, this is more a CMake usage question and not a libdatachannel question, I'd be better to learn how to use CMake or ask on a CMake help forum. You can also use any IDE for Windows which will take care of this for you.
